core/arch/x86_64/mm/core_mmu.c

```diff
@@ -821,6 +821,11 @@ static size_t collect_mem_ranges(struct tee_mmap_region *memory_map,
 		/* Only unmapped virtual range may have a null phys addr */
 		assert(m.addr || !core_mmu_type_to_attr(m.type));
 
+		if (m.type == MEM_AREA_IO_NSEC || m.type == MEM_AREA_IO_SEC) {
+			m.addr = ROUNDDOWN(m.addr, CORE_MMU_PGDIR_SIZE);
+			m.size = ROUNDUP(m.size + (mem->addr - m.addr),CORE_MMU_PGDIR_SIZE);
+		}
```

Comment:
Why do you need special handling for MEM_AREA_IO_NSEC  in collect_mem_ranges function?  Does Arm have the same code?